// @instantdb/admin/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tx = exports.id = exports.transact = exports.query = exports.init = void 0;
const instatx = __importStar(require("./instatx"));
const uuid = __importStar(require("uuid"));
let _GLOBAL_CONFIG = null;
function assert(condition, msg) {
    if (!condition) {
        throw new Error("[assertion error] " + msg);
    }
}
function authorizedHeaders(config) {
    const { adminToken, appId } = config;
    return {
        "content-type": "application/json",
        "authorization": `Bearer ${adminToken}`,
        "app-id": appId
    };
}
function jsonFetch(input, init) {
    return __awaiter(this, void 0, void 0, function* () {
        const res = yield fetch(input, init);
        const json = yield res.json();
        return res.status === 200
            ? Promise.resolve(json)
            : Promise.reject({ status: res.status, body: json });
    });
}
function getConfig() {
    assert(_GLOBAL_CONFIG, "Uh oh! Looks like `init` hasn't run yet. Make sure `init` runs " +
        "before your first `query` or `transact`.");
    return _GLOBAL_CONFIG;
}
function init(config) {
    _GLOBAL_CONFIG = config;
}
exports.init = init;
function query(query) {
    const config = getConfig();
    return jsonFetch(`${config.apiURI}/admin/query`, {
        method: "POST",
        headers: authorizedHeaders(config),
        body: JSON.stringify({ 'query': query }),
    });
}
exports.query = query;
function transact(inputChunks) {
    const config = getConfig();
    const chunks = Array.isArray(inputChunks) ? inputChunks : [inputChunks];
    const steps = chunks.flatMap((tx) => instatx.getOps(tx));
    return jsonFetch(`${config.apiURI}/admin/transact`, {
        method: "POST",
        headers: authorizedHeaders(config),
        body: JSON.stringify({ 'steps': steps }),
    });
}
exports.transact = transact;
exports.id = uuid.v4;
exports.tx = instatx.tx;
